File: drivers/ios/finalrun-ios-test/XCTest Models/ActionRequest.swift
import Foundation

// MARK: - Helper Types

struct Point: Codable {
    let x: Int
    let y: Int
}

struct PointPercent: Codable {
    let xP: Double
    let yP: Double
}

struct AppUpload: Codable {
    let packageName: String
}

struct SingleArgument: Codable {
    let type: String
    let value: String
}

// MARK: - Parent Action Enum

enum Action: Codable {
    case tap(TapAction)
    case tapPercent(TapPercentAction)
    case enterText(EnterTextAction)
    case scroll(ScrollAction)
    case eraseText(EraseTextAction)
    case copyText(CopyTextAction)
    case pasteText(PasteTextAction)
    case back(BackAction)
    case home(HomeAction)
    case rotate(RotateAction)
    case startStreaming(StartStreaming)
    case stopStreaming(StopStreaming)
    case stopExecution(StopExecution)
    case getHierarchy(GetHierarchy)
    case getScreenshot(GetScreenshot)
    case getScreenshotAndHierarchy(GetScreenshotAndHierarchy)
    case getScreenDimension(GetScreenDimension)
    case checkAppInForeground(CheckAppInForeground)
    case hideKeyboard(HideKeyboardAction)
    case killApp(KillAppAction)
    case launchApp(LaunchAppAction)
    case switchToPrimaryApp(SwitchToPrimaryAppAction)
    case pressKey(PressKeyAction)
    case setLocation(SetLocationAction)
    case updateAppIds(UpdateAppIdAction)
    case getDeviceScale(GetDeviceScaleAction)
    
    enum CodingKeys: String, CodingKey {
        case type
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        switch type {
        case "tap":
            self = .tap(try TapAction(from: decoder))
        case "tapPercent":
            self = .tapPercent(try TapPercentAction(from: decoder))
        case "enterText":
            self = .enterText(try EnterTextAction(from: decoder))
        case "scroll":
            self = .scroll(try ScrollAction(from: decoder))
        case "eraseText":
            self = .eraseText(try EraseTextAction(from: decoder))
        case "copyText":
            self = .copyText(try CopyTextAction(from: decoder))
        case "pasteText":
            self = .pasteText(try PasteTextAction(from: decoder))
        case "back":
            self = .back(try BackAction(from: decoder))
        case "home":
            self = .home(try HomeAction(from: decoder))
        case "rotate":
            self = .rotate(try RotateAction(from: decoder))
        case "startStreaming":
            self = .startStreaming(try StartStreaming(from: decoder))
        case "stopStreaming":
            self = .stopStreaming(try StopStreaming(from: decoder))
        case "stopExecution":
            self = .stopExecution(try StopExecution(from: decoder))
        case "getHierarchy":
            self = .getHierarchy(try GetHierarchy(from: decoder))
        case "getScreenshot":
            self = .getScreenshot(try GetScreenshot(from: decoder))
        case "getScreenshotAndHierarchy":
            self = .getScreenshotAndHierarchy(try GetScreenshotAndHierarchy(from: decoder))
        case "getScreenDimension":
            self = .getScreenDimension(try GetScreenDimension(from: decoder))
        case "checkAppInForeground":
            self = .checkAppInForeground(try CheckAppInForeground(from: decoder))
        case "hideKeyboard":
            self = .hideKeyboard(try HideKeyboardAction(from: decoder))
        case "killApp":
            self = .killApp(try KillAppAction(from: decoder))
        case "launchApp":
            self = .launchApp(try LaunchAppAction(from: decoder))
        case "switchToPrimaryApp":
            self = .switchToPrimaryApp(try SwitchToPrimaryAppAction(from: decoder))
        case "pressKey":
            self = .pressKey(try PressKeyAction(from: decoder))
        case "setLocation":
            self = .setLocation(try SetLocationAction(from: decoder))
        case "updateAppIds":
            self = .updateAppIds(try UpdateAppIdAction(from: decoder))
        case "getDeviceScale":
            self = .getDeviceScale(try GetDeviceScaleAction(from: decoder))
        default:
            throw DecodingError.dataCorruptedError(forKey: CodingKeys.type,
                                                   in: container,
                                                   debugDescription: "Unknown type: \(type)")
        }
    }
    
    func encode(to encoder: Encoder) throws {
        switch self {
        case .tap(let action):
            try action.encode(to: encoder)
        case .tapPercent(let action):
            try action.encode(to: encoder)
        case .enterText(let action):
            try action.encode(to: encoder)
        case .scroll(let action):
            try action.self.encode(to: encoder)
        case .eraseText(let action):
            try action.encode(to: encoder)
        case .copyText(let action):
            try action.encode(to: encoder)
        case .pasteText(let action):
            try action.encode(to: encoder)
        case .back(let action):
            try action.encode(to: encoder)
        case .home(let action):
            try action.encode(to: encoder)
        case .rotate(let action):
            try action.encode(to: encoder)
        case .startStreaming(let action):
            try action.encode(to: encoder)
        case .stopStreaming(let action):
            try action.encode(to: encoder)
        case .stopExecution(let action):
            try action.encode(to: encoder)
        case .getHierarchy(let action):
            try action.encode(to: encoder)
        case .getScreenshot(let action):
            try action.encode(to: encoder)
        case .getScreenshotAndHierarchy(let action):
            try action.encode(to: encoder)
        case .getScreenDimension(let action):
            try action.encode(to: encoder)
        case .checkAppInForeground(let action):
            try action.encode(to: encoder)
        case .hideKeyboard(let action):
            try action.encode(to: encoder)
        case .killApp(let action):
            try action.encode(to: encoder)
        case .launchApp(let action):
            try action.encode(to: encoder)
        case .switchToPrimaryApp(let action):
            try action.encode(to: encoder)
        case .pressKey(let action):
            try action.encode(to: encoder)
        case .setLocation(let action):
            try action.encode(to: encoder)
        case .updateAppIds(let action):
            try action.encode(to: encoder)
        case .getDeviceScale(let action):
            try action.encode(to: encoder)
        }
    }
}

// MARK: - Action Subclasses

struct TapAction: Codable {
    let point: Point
    let repeatCount: Int?
    let delay: Int?
    let requestId: String
    let type: String = "tap"
    
    enum CodingKeys: String, CodingKey {
        case point, requestId, delay
        case repeatCount = "repeat"
    }
}

struct TapPercentAction: Codable {
    let point: PointPercent
    let repeatCount: Int?
    let delay: Int?
    let requestId: String
    let type: String = "tapPercent"
    
    enum CodingKeys: String, CodingKey {
        case point, requestId, delay
        case repeatCount = "repeat"
    }
}

struct EnterTextAction: Codable {
    let value: String
    let requestId: String
    let shouldEraseText: Bool
    let type: String
    let eraseCount: Int?
    
    // Computed property to get erase count with default
    var effectiveEraseCount: Int {
        return eraseCount ?? 100
    }
}

struct ScrollAction: Codable {
    let requestId: String
    let type: String
    let duration: Int
    let x1: Int
    let x2: Int
    let y1: Int
    let y2: Int
}

struct EraseTextAction: Codable {
    let requestId: String
    let type: String
}

struct CopyTextAction: Codable {
    let requestId: String
    let type: String
}

struct PasteTextAction: Codable {
    let requestId: String
    let type: String
}

struct BackAction: Codable {
    let requestId: String
    let type: String
}

struct HomeAction: Codable {
    let requestId: String
    let type: String
}

struct RotateAction: Codable {
    let requestId: String
    let type: String
}

struct StartStreaming: Codable {
    let requestId: String
    let fps: Int?   // Default can be provided by your business logic (e.g., 24)
    let quality: Int? // Default can be provided (e.g., 5)
    let type: String
}

struct StopStreaming: Codable {
    let requestId: String
    let type: String
}

struct StopExecution: Codable {
    let requestId: String
    let type: String
}

struct GetHierarchy: Codable {
    let requestId: String
    let type: String
}

struct GetScreenshot: Codable {
    let requestId: String
    let type: String
    let quality: Int?
}

struct GetScreenshotAndHierarchy: Codable {
    let requestId: String
    let type: String
    let quality: Int?
}

struct GetScreenDimension: Codable {
    let requestId: String
    let type: String
}

struct CheckAppInForeground: Codable {
    let requestId: String
    let packageName: String
    let timeout: Int  // timeout in seconds
    let type: String
}

struct HideKeyboardAction: Codable {
    let requestId: String
    let type: String
}

struct KillAppAction: Codable {
    let requestId: String
    let type: String
}

struct LaunchAppAction: Codable {
    let appUpload: AppUpload
    let allowAllPermissions: Bool
    let arguments: [String: SingleArgument]?
    let permissions: [String: String]?
    let shouldUninstallBeforeLaunch: Bool
    let requestId: String
    let type: String
}

struct SwitchToPrimaryAppAction: Codable {
    let packageName: String
    let requestId: String
    let type: String
}

struct PressKeyAction: Codable {
    let key: String
    let requestId: String
    let type: String
}

struct SetLocationAction: Codable {
    let lat: Double
    let long: Double
    let requestId: String
    let type: String
}

struct UpdateAppIdAction: Codable {
    let appIds: [String]
    let requestId: String
    let type: String
}

struct GetDeviceScaleAction: Codable {
    let requestId: String
    let type: String
}

File: drivers/android/app/src/androidTest/java/app/finalrun/android/data/ActionRequest.kt
package app.finalrun.android.data

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.annotation.JsonTypeName
import com.fasterxml.jackson.annotation.JsonProperty

// ----- Helper classes (stubs) -----
data class Point(
    @JsonProperty("x") val x: Int,
    @JsonProperty("y") val y: Int
)

data class PointPercent(
    @JsonProperty("xP") val xP: Double,
    @JsonProperty("yP") val yP: Double
)

data class AppUpload(
    @JsonProperty("packageName") val packageName: String
)

data class SingleArgument(
    @JsonProperty("type") val type: String,
    @JsonProperty("value") val value: String
)

// ----- Base Action class with Jackson annotations -----
@JsonTypeInfo(
    use = JsonTypeInfo.Id.NAME,
    include = JsonTypeInfo.As.PROPERTY,
    property = "type"
)
@JsonSubTypes(
    JsonSubTypes.Type(value = TapAction::class, name = Action.TAP),
    JsonSubTypes.Type(value = TapPercentAction::class, name = Action.TAP_PERCENT),
    JsonSubTypes.Type(value = EnterTextAction::class, name = Action.ENTER_TEXT),
    JsonSubTypes.Type(value = EraseTextAction::class, name = Action.ERASE_TEXT),
    JsonSubTypes.Type(value = CopyTextAction::class, name = Action.COPY_TEXT),
    JsonSubTypes.Type(value = PasteTextAction::class, name = Action.PASTE_TEXT),
    JsonSubTypes.Type(value = BackAction::class, name = Action.BACK),
    JsonSubTypes.Type(value = HomeAction::class, name = Action.HOME),
    JsonSubTypes.Type(value = RotateAction::class, name = Action.ROTATE),
    JsonSubTypes.Type(value = HideKeyboardAction::class, name = Action.HIDE_KEYBOARD),
    JsonSubTypes.Type(value = KillAppAction::class, name = Action.KILL_APP),
    JsonSubTypes.Type(value = LaunchAppAction::class, name = Action.LAUNCH_APP),
    JsonSubTypes.Type(value = PressKeyAction::class, name = Action.PRESS_KEY),
    JsonSubTypes.Type(value = SetLocationAction::class, name = Action.SET_LOCATION),
    JsonSubTypes.Type(value = GetAppListAction::class, name = Action.GET_APP_LIST),
    JsonSubTypes.Type(value = DeviceScaleAction::class, name = Action.GET_DEVICE_SCALE),
)
sealed class Action(@JsonProperty("requestId") open val requestId: String) {
    abstract val type: String

    companion object {
        const val TAP = "tap"
        const val TAP_PERCENT = "tapPercent"
        const val ENTER_TEXT = "enterText"
        const val ERASE_TEXT = "eraseText"
        const val COPY_TEXT = "copyText"
        const val PASTE_TEXT = "pasteText"
        const val BACK = "back"
        const val HOME = "home"
        const val ROTATE = "rotate"
        const val START_STREAMING = "startStreaming"
        const val STOP_STREAMING = "stopStreaming"
        const val STOP_EXECUTION = "stopExecution"
        const val GET_HIERARCHY = "getHierarchy"
        const val GET_SCREENSHOT = "getScreenshot"
        const val GET_SCREENSHOT_AND_HIERARCHY = "getScreenshotAndHierarchy"
        const val GET_SCREEN_DIMENSION = "getScreenDimension"
        const val HIDE_KEYBOARD = "hideKeyboard"
        const val KILL_APP = "killApp"
        const val LAUNCH_APP = "launchApp"
        const val SWITCH_TO_PRIMARY_APP = "switchToPrimaryApp"
        const val PRESS_KEY = "pressKey"
        const val GET_DEVICE_SCALE = "getDeviceScale"
        const val SET_LOCATION = "setLocation"
        const val CHECK_APP_IN_FOREGROUND = "checkAppInForeground"
        const val GET_APP_LIST = "getAppList"
    }
}

// ----- Action Subclasses -----

@JsonTypeName(Action.COPY_TEXT)
data class CopyTextAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = COPY_TEXT
}

@JsonTypeName(Action.PASTE_TEXT)
data class PasteTextAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = PASTE_TEXT
}

@JsonTypeName(Action.TAP)
data class TapAction(
    @JsonProperty("point") val point: Point,
    @JsonProperty("repeat") val repeat: Int? = null,
    @JsonProperty("delay") val delay: Int? = null,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = TAP
}

@JsonTypeName(Action.TAP_PERCENT)
data class TapPercentAction(
    @JsonProperty("point") val point: PointPercent,
    @JsonProperty("repeat") val repeat: Int? = null,
    @JsonProperty("delay") val delay: Int? = null,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = TAP_PERCENT
}

@JsonTypeName(Action.ENTER_TEXT)
data class EnterTextAction(
    @JsonProperty("value") val value: String,
    @JsonProperty("requestId") override val requestId: String,
    @JsonProperty("shouldEraseText") val shouldEraseText: Boolean,
    @JsonProperty("eraseCount") val eraseCount: Int? = null,  // Optional, used for erasing text
) : Action(requestId) {
    override val type: String = ENTER_TEXT
}

@JsonTypeName(Action.ERASE_TEXT)
data class EraseTextAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = ERASE_TEXT
}

@JsonTypeName(Action.BACK)
data class BackAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = BACK
}

@JsonTypeName(Action.GET_SCREENSHOT)
data class GetScreenshot(
    @JsonProperty("requestId") override val requestId: String,
    @JsonProperty("quality") val quality: Int? = 5,
) : Action(requestId) {
    override val type: String = GET_SCREENSHOT
}

@JsonTypeName(Action.GET_HIERARCHY)
data class GetHierarchy(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = GET_HIERARCHY
}

@JsonTypeName(Action.GET_SCREENSHOT_AND_HIERARCHY)
data class GetScreenshotAndHierarchy(
    @JsonProperty("requestId") override val requestId: String,
    @JsonProperty("quality") val quality: Int? = 5,
) : Action(requestId) {
    override val type: String = GET_SCREENSHOT_AND_HIERARCHY
}

@JsonTypeName(Action.GET_SCREEN_DIMENSION)
data class GetScreenDimension(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = GET_SCREEN_DIMENSION
}

@JsonTypeName(Action.CHECK_APP_IN_FOREGROUND)
data class CheckAppInForeground(
    @JsonProperty("requestId") override val requestId: String,
    @JsonProperty("packageName") val packageName: String,
    @JsonProperty("timeout") val timeout: Int // timeout in seconds
) : Action(requestId) {
    override val type: String = CHECK_APP_IN_FOREGROUND
}

@JsonTypeName(Action.HOME)
data class HomeAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = HOME
}

@JsonTypeName(Action.ROTATE)
data class RotateAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = ROTATE
}

@JsonTypeName(Action.START_STREAMING)
data class StartStreaming(
    @JsonProperty("requestId") override val requestId: String,
    @JsonProperty("fps") val fps: Int? = 24,
    @JsonProperty("quality") val quality: Int? = 5,
) : Action(requestId) {
    override val type: String = START_STREAMING
}

@JsonTypeName(Action.STOP_STREAMING)
data class StopStreaming(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = STOP_STREAMING
}

@JsonTypeName(Action.STOP_EXECUTION)
data class StopExecution(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = STOP_EXECUTION
}

@JsonTypeName(Action.HIDE_KEYBOARD)
data class HideKeyboardAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = HIDE_KEYBOARD
}

@JsonTypeName(Action.KILL_APP)
data class KillAppAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = KILL_APP
}

@JsonTypeName(Action.LAUNCH_APP)
data class LaunchAppAction(
    @JsonProperty("appUpload") val appUpload: AppUpload,
    @JsonProperty("allowAllPermissions") val allowAllPermissions: Boolean,
    @JsonProperty("arguments") val arguments: Map<String, SingleArgument>? = null,
    @JsonProperty("permissions") val permissions: Map<String, String>? = null,
    @JsonProperty("shouldUninstallBeforeLaunch") val shouldUninstallBeforeLaunch: Boolean,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = LAUNCH_APP
}

@JsonTypeName(Action.SWITCH_TO_PRIMARY_APP)
data class SwitchToPrimaryAppAction(
    @JsonProperty("packageName") val packageName: String,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = SWITCH_TO_PRIMARY_APP
}

@JsonTypeName(Action.PRESS_KEY)
data class PressKeyAction(
    @JsonProperty("key") val key: String,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = PRESS_KEY
}

@JsonTypeName(Action.GET_DEVICE_SCALE)
data class DeviceScaleAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = GET_DEVICE_SCALE
}

@JsonTypeName(Action.SET_LOCATION)
data class SetLocationAction(
    @JsonProperty("lat") val lat: Double,
    @JsonProperty("long") val long: Double,
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = SET_LOCATION
}

@JsonTypeName(Action.GET_APP_LIST)
data class GetAppListAction(
    @JsonProperty("requestId") override val requestId: String
) : Action(requestId) {
    override val type: String = GET_APP_LIST
}
